### PR TITLE
TST: add benchmarking and Faker support

### PR DESCRIPTION
**Pull request by aqw**
The core motivation for this PR is to add some benchmarks ~~and test out CodSpeed to see if it's a reasonable fit for us~~.

~~If not, the benchmarks are still useful, but we'll go back to vanilla pytest-benchmark. Codspeed is mostly compatible, until it isn't. For example, it does not support `pendantic`, which is a loss.~~

Generating large, representative repos led me to finally add Faker support. I hope/plan to use this for more tests that require populated repos, rather than our current verbose, hard-coded mess.

This PR also:
- fixes a bug in `git.get_hexsha()`
- fixes a documentation bug WRT flake8->ruff
- normalizes `onyo_cat()`'s arguments to match other onyo functions.

Fixes:
- #224 
- #254

I encountered numerous bugs around pytest fixture scoping, parametrization, indirect parametrization, and referring to the parent class.

As a result, I hack a class-level cache through a function-scoped fixture. I would like to to this differently, with a class-level fixture populating and function-level fixture cloning, but this was not to be.

I also setup Faker with fixtures for each scope. We may not use this, in which case we can remove it. But I wanted to include it for now so we have a history of it in the repo, and it may in fact turn out to be of use.

### CODE REVIEW

File: onyo/conftest.py [hunk 1]
```diff
@@ -258,3 +259,160 @@ def set_ui(request) -> None:
...
+
+class OnyoProvider(python.Provider):
+    r"""
+    Faker Provider for Onyo.
+    """
+    def onyo_asset_dicts(self,
+                         num: int = 1,
+                         override: dict | None = None) -> Generator[dict, None, None]:
+        r"""
+        Yield asset directories suitable for populating realistic assets.
```
Comment (bpoldrack) [inline onyo/conftest.py:272]:
```suggestion
        Yield asset dictionaries suitable for populating realistic assets.
```

**Review by bpoldrack — APPROVED** [review]
Apart from the benchmark build hanging, this is looking good to me. Thanks much!

**Comment by aqw** [comment]
I have pushed a new set of commits to this PR.

Primary changes are:
- faker is now available across multiple scopes
- drop CodSpeed and use pytest-benchmark directly (soooo much nicer)
- Add a class-scoped ui marker

If there are any questions on my approach, I am happy to explain. I'm not thrilled with the solution, but it was the best I could manage given the numerous pytest bugs I encountered.

**Review by bpoldrack — APPROVED** [review]
Thanks much, @aqw !

File: onyo/conftest.py [hunk 2]
```diff
@@ -250,11 +252,207 @@ def helpers() -> Type[Helpers]:
...
+        )
+        yield self.random_element(elements=cpu_vendors)
+
+
+    def onyo_directories(self,
+                         num: int = 1) -> Generator[str, None, None]:
+        r"""
+        Yield directory names.
+
+        Useful for ``onyo.path.directory``.
```
[the remark this answers is not in the document]
Comment (aqw) [reply onyo/conftest.py:344]: It is a reference to future (near imminent) pseudo-keys.

IMO, it can stay as is, but if you'd like, I can change it.

**Comment by aqw** [comment]
Thanks for the review!

**Pull request merged**